File: basic/java/search/BinarySearch.java
package java.search;

// only for sorted array
//
// time complexity  : O(log n)
// space complexity : O(1)
//
// Divide and conquer

public class BinarySearch {
    public int binarySearch(int[] arr, int target) {
        int low = 0;
        int high = arr.length - 1; 
  
        while (low <= high) {
           int mid = low + (high - low) / 2; 
  
           if (arr[mid] == target) {
              return mid;   // target found, return index
           }

           if (target < arr[mid]) {
              high = mid - 1;
           } else {
              low = mid + 1;
           }
        }
        
        return -1;  // target not found
        // alternative: if we wanted to know where a target should be placed, we could return "low"
     }

     public static void main(String[] args) {
        BinarySearch bs = new BinarySearch();
        int[] arr = { 1, 10, 20, 47, 59, 65, 75, 88, 99 };
        System.out.println(bs.binarySearch(arr, 65));
     }
    
}

File: basic/python/search/binarySearch.py
# only on sorted list
# time complexity: O(log n) 
# Divide and conquer

# Iterative Binary Search
def binary_search(list, target):
    low = 0
    high = len(list) - 1

    while low <= high:
        mid = low + (high - low) // 2   # Find the middle index using integer division, and avoid overflow
        
        if list[mid] == target:
            return mid            # Target found
        elif list[mid] < target:
            low = mid + 1         # Discard the left half
        else:
            high = mid - 1        # Discard the right half
            
    return None  # Target not found

 
# Recursive Binary Search
def recursive_binary_search(list, target):
    if len(list) == 0:
        return False
    else:
        mid = (len(list)) // 2
        if list[mid] == target:
            return True
        else:
            if list[mid] < target:
                return recursive_binary_search(list[mid+1:], target)
            else:
                return recursive_binary_search(list[:mid], target)
            
